RedirectSOAuthorizationSession should support all redirections

youennf opened the pull request:
#### 996faadd2f4f22a27bb80545a14a36d5007ae466
<pre>
RedirectSOAuthorizationSession should support all redirections
<a href="https://bugs.webkit.org/show_bug.cgi?id=254263">https://bugs.webkit.org/show_bug.cgi?id=254263</a>
rdar://106379935

Reviewed by Alex Christensen.

Handle SSO redirections in the loader code path.
This has the benefit of supporting all redirections, including 307 redirections.
This also simplifies the UIProcess code path.

We need to handle redirections in two ways:
- Either the SSO redirection happens while the load has not started.
  We handle it at DocumentLoader level by enabling to use substitute data for redirections.
- Or the SSO redirection happens while the load already started (redirection case).
  In that case, we register the redirection in network process so that NetworkResourceLoader handles it.

We add some tests and need to update some tests as redirections from file URLs to HTTP is not allowed.

This new code path is only enabled for 307 redirections on a POST request for now.
This patch should not change behavior for other interceptions.
In the future, we will probably move all redirections to follow that new code path.

Coverd by existing and added API tests.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::handleSubstituteDataLoadNow):
(WebCore::DocumentLoader::redirectReceived):
(WebCore::DocumentLoader::setRedirectionAsSubstituteData):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::useRedirectionForCurrentNavigation):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::continueWillSendRequest):
(WebKit::NetworkResourceLoader::useRedirectionForCurrentNavigation):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/RedirectSOAuthorizationSession.mm:
(WebKit::RedirectSOAuthorizationSession::completeInternal):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::useRedirectionForCurrentNavigation):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::useRedirectionForCurrentNavigation):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SOAuthorizationTests.mm:
(-[TestSOAuthorizationBasicDelegate webView:didFinishNavigation:]):
(-[TestSOAuthorizationDelegate webView:didFinishNavigation:]):
(-[TestSOAuthorizationDelegate webView:decidePolicyForNavigationAction:decisionHandler:]):
(overrideBeginAuthorizationWithURL):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/265706@main">https://commits.webkit.org/265706@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/913fff5b72f26c81501375ed3e672a50628ea615

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11692 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11892 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12259 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13336 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11137 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11711 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14280 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11872 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14003 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11856 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12697 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9919 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13757 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9989 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10615 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/17749 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11066 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10769 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13943 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11169 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9217 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10346 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2804 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14626 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11026 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
```diff
@@ -420,29 +420,7 @@ static void tryInterceptNavigation(Ref<API::NavigationAction>&& navigationAction
                 return;
             }
 
-            if (!navigationAction->targetFrame()) {
```
Why do you think it's safe to remove all of this stuff? It doesn't look like this was added to support AppSSO, so it seems unnecessarily risky to remove this logic without a clear reason.

brentfulgham replied:
If you think it's just dead code, I would suggest creating a separate PR for that so we can easily revert the change if we discover it breaks something.

youennf replied:
OK, I'll add this code back and we can remove it as a follow-up if we find out this is fine.

youennf replied:
Added back the code.

brentfulgham commented on Source/WebKit/UIProcess/Cocoa/SOAuthorization/RedirectSOAuthorizationSession.mm:
```diff
@@ -67,34 +67,22 @@
-        // workaround this issue, we load an html page that does a client side redirection to the application site on behalf of the
-        // request URL, instead of directly loading a new request. The html page should be super fast to load and therefore will not
-        // show an empty URL or a blank page. These changes ensure a relevant URL bar and useful page content during the load.
-        if (!navigationAction->isProcessingUserGesture()) {
```
Your comments don't explain why you feel that it is safe to remove this logic. I guess the idea is that we don't fire the redirect at all anymore, we just register the redirect then proceed so that the redirect will get handled as the next step. So therefore we don't need to invoke the spoof mitigation.

I'm worried we still need the Spoof mitigation; I wonder if you could build the data URL in a new request, and register *that* before failing the callback?

youennf replied:
I am not sure either how to test it.
What happens now with the patch is as if the load proceeds and there is a server side redirection.
The spoofing mitigator should already be handling correctly server side redirections (probably not do anything).

brentfulgham reviewed (commented):
I don't feel confident about the two large blocks of code you are removing. I think the overall direction of this patch is good, but I worry that the spoof mitigation is still needed (and I'm not sure how to test that).

I also worry that your change removes code needed to support some kinds of downloads. How can we confirm the change is safe?

shivamidow commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8728,6 +8729,33 @@ bool WebPage::shouldSkipDecidePolicyForResponse(const WebCore::ResourceResponse&
+        return;
+    }
+
+    auto* loader = localMainFrame ? localMainFrame->loader().policyDocumentLoader() : nullptr;
```
/* Unofficial Review */
localMainFrame should not be a nullptr here and below. It is checked above.
auto* loader = localMainFrame->loader().policyDocumentLoader() is enough.

youennf replied:
done

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SOAuthorizationTests.mm:
```diff
@@ -54,6 +54,7 @@
 static bool uiShowed = false;
 static bool newWindowCreated = false;
 static bool haveHttpBody = false;
+static RetainPtr<NSData> httpBody;
```
This should be an ivar of the delegate

youennf replied:
Not really, the SOA plugin is using static functions

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
```diff
@@ -721,5 +721,7 @@ GenerateSyntheticEditingCommand(enum:uint8_t WebKit::SyntheticEditingCommandType
     PlayAllAnimations() -> ()
 #endif
 
+    RegisterRedirectionToFollow(WebCore::ResourceResponse response);
```
This is a poor name.  Just looking at this message I have no idea what it is intended to be used for, what the response represents, etc.

youennf replied:
Changed to useRedirectionForCurrentNavigation

achristensen07 commented on Source/WebCore/loader/SubstituteData.h:
```diff
@@ -51,7 +51,7 @@ class SubstituteData {
     }
 
-    bool isValid() const { return m_content != nullptr; }
+    bool isValid() const { return m_content != nullptr || m_response.isRedirection(); }
```
This seems suspicious.  Why would a redirection response but no content be a valid substitute data?

youennf replied:
It should be fine, a redirection does not need content, but I removed this change as I can do otherwise.

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1465,6 +1465,12 @@ void NetworkConnectionToWebProcess::logOnBehalfOfWebContent(const String& logCha
 
+void NetworkConnectionToWebProcess::registerRedirectionToFollow(WebCore::ResourceLoaderIdentifier identifier, WebCore::ResourceResponse&& response)
+{
+    if (RefPtr<NetworkResourceLoader> loader = m_networkResourceLoaders.get(identifier))
```
<NetworkResourceLoader> is unnecessary.  Might need { }

youennf replied:
done

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/SOAuthorization/RedirectSOAuthorizationSession.mm:
```diff
@@ -67,34 +67,22 @@
         return;
     }
-    invokeCallback(true);
-    if (response.httpStatusCode() == 302) {
```
I would expect this PR to only change this from ==302 to isRedirection()
Why are we moving around so many things?

youennf replied:
The issue is that previously, when receiving the 301, we were simulating the 301 redirection by triggering a navigation to the end URL. This was a quick and easy way to get the job done.
This is not as easy with POST redirections, given we do not have the body.
And this hack is also triggering issues with Safari as Safari is not understanding this is a redirection.

We are now handling the redirection in web process, as if it was a redirection coming from networking.
This is the more legit way of doing things I think.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SOAuthorizationTests.mm:
```diff
@@ -740,6 +769,188 @@ static void checkAuthorizationOptions(bool userActionInitiated, String initiator
+    [webView loadRequest:[NSURLRequest requestWithURL:[server.request() URL]]];
+    Util::run(&navigationCompleted);
+
+    haveHttpBody = true;
```
Do we check anywhere that the HTTP body is actually sent as expected?

youennf replied:
We check the navigation action delegate gets a POST with  the data to the redirected URL.
We do not check that the server itself gets the POST. We could add that if there is a way to do so.

achristensen07 replied:
Look for UseCoroutines.  That kind of test allows you to inspect each request from the server.

youennf replied:
Done

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SOAuthorizationTests.mm:
```diff
@@ -646,11 +667,15 @@ static void checkAuthorizationOptions(bool userActionInitiated, String initiator
-    auto response = adoptNS([[NSHTTPURLResponse alloc] initWithURL:testURL statusCode:302 HTTPVersion:@"HTTP/1.1" headerFields:@{ @"Location" : [redirectURL absoluteString] }]);
+    TestWebKitAPI::HTTPServer server({
+        { "/"_s, { SimpleHtml } },
+    }, TestWebKitAPI::HTTPServer::Protocol::Http, nullptr, nullptr, 8091);
```
Why port 8091?  Is that just one you're guessing is available?  It's typically better to let the server pick one that is available and send requests to whichever one it picks.

youennf replied:
Done

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SOAuthorizationTests.mm:
```diff
@@ -740,6 +769,248 @@ static void checkAuthorizationOptions(bool userActionInitiated, String initiator
+        }
+    }, TestWebKitAPI::HTTPServer::Protocol::Http, 8091);
+
+    [webView loadRequest:[NSURLRequest requestWithURL:[server.request() URL]]];
```
This is redundant.  Just use server.request instead of extracting the URL and putting it back into a request.

youennf replied:
Done

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SOAuthorizationTests.mm:
```diff
@@ -740,6 +769,248 @@ static void checkAuthorizationOptions(bool userActionInitiated, String initiator
+
+    EXPECT_EQ(WKNavigationTypeOther, [delegateNavigationAction navigationType]);
+    EXPECT_WK_STREQ("GET", [[delegateNavigationAction request] HTTPMethod]);
+    EXPECT_WK_STREQ("http://127.0.0.1:8091/simple2.html", [[[delegateNavigationAction request] URL] absoluteString]);
```
I think dot syntax would be nicer here

youennf replied:
Switched to dot.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SOAuthorizationTests.mm:
```diff
@@ -740,6 +769,248 @@ static void checkAuthorizationOptions(bool userActionInitiated, String initiator
+
+    haveHttpBody = true;
+
+    RetainPtr<NSString> formSubmissionScript = [NSString stringWithFormat:@"submitForm('<form action=\"%@\" method=\"POST\"><input type=\"text\" name=\"name\" id=\"name\" value=\"test\"></input><input type=\"submit\" name=\"submitButton\" value=\"Submit\"></form>')", [[server.request("/simple.html"_s) URL] absoluteString]];
```
The absoluteString call isn't necessary here.

youennf replied:
Done

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SOAuthorizationTests.mm:
```diff
@@ -740,6 +769,248 @@ static void checkAuthorizationOptions(bool userActionInitiated, String initiator
+    EXPECT_TRUE([delegateNavigationAction sourceFrame] == [delegateNavigationAction targetFrame]);
+    EXPECT_WK_STREQ("POST", [[delegateNavigationAction request] HTTPMethod]);
+    EXPECT_WK_STREQ("http://127.0.0.1:8091/simple2.html", [[[delegateNavigationAction request] URL] absoluteString]);
+    EXPECT_WK_STREQ(adoptNS([[NSString alloc] initWithData:[[delegateNavigationAction request] HTTPBody] encoding:NSUTF8StringEncoding]).get(), "name=test");
```
Write a helper function that checks HTTP data.  This is a lot of redundant code.

achristensen07 commented on Tools/TestWebKitAPI/cocoa/HTTPServer.h:
```diff
@@ -89,7 +89,7 @@ class HTTPServer {
     HTTPServer(Function<void(Connection)>&&, Protocol = Protocol::Http);
     enum class UseCoroutines : bool { Yes };
-    HTTPServer(UseCoroutines, Function<Task(Connection)>&&, Protocol = Protocol::Http);
+    HTTPServer(UseCoroutines, Function<Task(Connection)>&&, Protocol = Protocol::Http, std::optional<uint16_t> port = { });
```
This isn't a bad change, but I don't see any good use of it in this PR.

youennf replied:
Removed

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/SOAuthorization/RedirectSOAuthorizationSession.mm:
```diff
@@ -67,13 +67,15 @@
     ASSERT(navigationAction);
     auto* page = this->page();
-    if ((response.httpStatusCode() != 302 && response.httpStatusCode() != 200) || !page) {
+    // FIXME: Enable the useRedirectionForCurrentNavigation code path for all redirections.
```
This looks a little strange.  Are we planning to do this in two steps?

youennf replied:
Yes, we should do it for all 30X redirections, no matter GET or POST.
This will allow removing some iOS specific code for instance.
I limit the changes to 307 POST for now to fix the most urgent issue.

achristensen07 commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -4359,5 +4359,3 @@ webaudio/ScriptProcessor/scriptprocessor-offlineaudiocontext.html [ Failure ]
 
 webkit.org/256217 [ Debug ] imported/w3c/web-platform-tests/background-fetch/port-blocking.https.window.html [ Pass Crash ]
-
-webkit.org/b/258328 [ Debug ] imported/w3c/web-platform-tests/webcodecs/videoFrame-construction.crossOriginSource.sub.html [ Pass Failure Crash ]
```
I don't understand how this is related.

youennf replied:
This is unrelated, I probably did a bad rebase.

youennf committed:
RedirectSOAuthorizationSession should support all redirections
https://bugs.webkit.org/show_bug.cgi?id=254263
rdar://106379935

Reviewed by Alex Christensen.

Handle SSO redirections in the loader code path.
This has the benefit of supporting all redirections, including 307 redirections.
This also simplifies the UIProcess code path.

We need to handle redirections in two ways:
- Either the SSO redirection happens while the load has not started.
  We handle it at DocumentLoader level by enabling to use substitute data for redirections.
- Or the SSO redirection happens while the load already started (redirection case).
  In that case, we register the redirection in network process so that NetworkResourceLoader handles it.

We add some tests and need to update some tests as redirections from file URLs to HTTP is not allowed.

This new code path is only enabled for 307 redirections on a POST request for now.
This patch should not change behavior for other interceptions.
In the future, we will probably move all redirections to follow that new code path.

Coverd by existing and added API tests.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::handleSubstituteDataLoadNow):
(WebCore::DocumentLoader::redirectReceived):
(WebCore::DocumentLoader::setRedirectionAsSubstituteData):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::useRedirectionForCurrentNavigation):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::continueWillSendRequest):
(WebKit::NetworkResourceLoader::useRedirectionForCurrentNavigation):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/RedirectSOAuthorizationSession.mm:
(WebKit::RedirectSOAuthorizationSession::completeInternal):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::useRedirectionForCurrentNavigation):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::useRedirectionForCurrentNavigation):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SOAuthorizationTests.mm:
(-[TestSOAuthorizationBasicDelegate webView:didFinishNavigation:]):
(-[TestSOAuthorizationDelegate webView:didFinishNavigation:]):
(-[TestSOAuthorizationDelegate webView:decidePolicyForNavigationAction:decisionHandler:]):
(overrideBeginAuthorizationWithURL):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/265706@main
